ateam_kenobi/src/path_planning/path_planner.cpp

```diff
@@ -0,0 +1,233 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#include "path_planner.hpp"
+#include <ranges>
+#include <algorithm>
+#include <vector>
+#include <ateam_common/robot_constants.hpp>
+#include <ateam_geometry/ateam_geometry.hpp>
+
+namespace ateam_kenobi::path_planning
+{
+
+PathPlanner::PathPlanner()
+{
+}
+
+PathPlanner::Path PathPlanner::getPath(
+  const Position & start, const Position & goal, const World & world,
+  const std::vector<ateam_geometry::AnyShape> & obstacles,
+  const PlannerOptions & options)
+{
+  const auto start_time = std::chrono::steady_clock::now();
+
+  std::vector<ateam_geometry::AnyShape> augmented_obstacles = obstacles;
+
+  addRobotsToObstacles(world, start, augmented_obstacles);
+
+  if (options.use_default_obstacles) {
+    addDefaultObstacles(world, augmented_obstacles);
+  }
+
+  if (options.avoid_ball) {
+    augmented_obstacles.push_back(ateam_geometry::makeCircle(world.ball.pos, 0.04267 / 2));
+  }
+
+  // if (!isStateValid(start, world, augmented_obstacles, options)) {
+  //   return {};
+  // }
+  if (!isStateValid(goal, world, augmented_obstacles, options)) {
+    return {};
+  }
+
+  Path path = {start, goal};
+
+  while (true) {
+    const auto elapsed_time = std::chrono::duration_cast<std::chrono::duration<double>>(
+      std::chrono::steady_clock::now() - start_time).count();
+    if (elapsed_time > options.search_time_limit) {
+      return path;
+    }
+    bool had_to_split = false;
+    for (auto ind = 0u; ind < (path.size() - 1); ++ind) {
+      auto split_result = splitSegmentIfNecessary(
+        path, ind, ind + 1, world, augmented_obstacles,
+        options);
+      if (split_result.split_needed) {
+        had_to_split = true;
+        if (!split_result.split_succeeded) {
+          // unable to find path around obstacle
+          return {};
+        }
+        break;
+      }
+    }
+    if (!had_to_split) {
+      // no split -> path is collision free
+      break;
+    }
+  }
+
+  return path;
+}
+
+bool PathPlanner::isStateValid(
+  const ateam_geometry::Point & state,
+  const World & world,
+  const std::vector<ateam_geometry::AnyShape> & obstacles,
+  const PlannerOptions & options)
+{
+  const auto x_bound = (world.field.field_length / 2.0) + world.field.boundary_width - kRobotRadius;
+  const auto y_bound = (world.field.field_width / 2.0) + world.field.boundary_width - kRobotRadius;
+  const ateam_geometry::Rectangle pathable_region(
+    ateam_geometry::Point(-x_bound, -y_bound),
+    ateam_geometry::Point(x_bound, y_bound));
```

Comment:
We end up calling this function a lot, and this value doesn't change. Might be better to pull this out as a class member instead of recalculating it every time we want to see if a point is valid.

Reply:
Just need to be careful about the "this doesn't change part" because I believe this comes from vision, so it will probably change at least once when we start getting field geometry packets.